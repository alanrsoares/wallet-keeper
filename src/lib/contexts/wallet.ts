import { createContainer } from "unstated-next";

import usePersistedState from "~/lib/hooks/usePersistedState";

export const STORAGE_KEY = "@walletkeeper/state/v1";

const { Provider: WalletProvider, useContainer: useWallet } = createContainer(
  () => {
    const [state, setState] = usePersistedState(STORAGE_KEY, {});

    return {
      state,
      setState,
    };
  }
);

export { WalletProvider, useWallet };
