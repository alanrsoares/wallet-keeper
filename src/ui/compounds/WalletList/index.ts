export { default } from "./WalletList";
export * from "./WalletList";
