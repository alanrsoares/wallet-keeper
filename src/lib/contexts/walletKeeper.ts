import type { ProgressCallback } from "@ethersproject/json-wallets";
import type { Networkish } from "@ethersproject/networks";
import { getDefaultProvider, Provider } from "@ethersproject/providers";
import { formatEther } from "@ethersproject/units";
import { Wallet } from "@ethersproject/wallet";
import { useMutation, useQuery } from "@tanstack/react-query";
import { getAddress } from "ethers/lib/utils";
import { useMemo } from "react";
import { createContainer } from "unstated-next";

import usePersistedState from "~/lib/hooks/usePersistedState";
import { isValidBIP32Address, sanitizeWalletName } from "../utils";

export const STORAGE_KEY = "@walletkeeper/state/v1";

export const NETWORKS = {
  mainnet: "homestead",
  goerli: "goerli",
};

const ALCHEMY_API_KEY = process.env.NEXT_PUBLIC_ALCHEMY_API_KEY;

export type AccountEntry = {
  address: string;
  encryptedJson: string;
  displayName: string;
};

export type GetBalanceParams = {
  address: string;
  provider?: Provider;
};

export type CreateWalletInput = {
  displayName: string;
  password: string;
  onProgress?: ProgressCallback;
};

export type ImportWalletInput = {
  displayName: string;
  privateKey: string;
  password: string;
  onProgress?: ProgressCallback;
};

export type RenameWalletInput = {
  displayName: string;
  address: string;
};

export type UnlockWalletInput = {
  address: string;
  password: string;
  onProgress?: ProgressCallback;
};

export type DeleteWalletInput = {
  address: string;
  password: string;
  onProgress?: ProgressCallback;
};

export type WalletKeeperState = {
  /**
   * Accounts, indexed by address
   */
  accountsByAddress: Record<string, AccountEntry>;
  /**
   * Selected network
   * @default "goerli"
   **/
  selectedNetwork?: Networkish;
};

const DEFAULT_STATE: WalletKeeperState = {
  accountsByAddress: {},
  selectedNetwork: NETWORKS.goerli,
};

const { Provider: WalletKeeperProvider, useContainer: useWalletKeeper } =
  createContainer((initialState: WalletKeeperState = DEFAULT_STATE) => {
    const [state, setState] = usePersistedState<WalletKeeperState>(
      STORAGE_KEY,
      initialState
    );

    const mutations = {
      generateWallet: () => {
        async function generateWalletMutation(input: CreateWalletInput) {
          const existingWallet = Object.values(state.accountsByAddress).find(
            (account) => account.displayName === input.displayName
          );

          if (existingWallet) {
            throw new Error("Wallet with that name already exists");
          }

          if (!input.password.length) {
            throw new Error("Password must not be empty");
          }

          const wallet = Wallet.createRandom();
          const encryptedJson = await wallet.encrypt(
            input.password,
            {},
            input.onProgress
          );

          setState((state) => ({
            ...state,
            accountsByAddress: {
              ...state.accountsByAddress,
              [wallet.address]: {
                address: wallet.address,
                encryptedJson,
                displayName: sanitizeWalletName(input.displayName),
              },
            },
          }));

          return wallet;
        }

        return useMutation(generateWalletMutation);
      },

      renameWallet: () => {
        async function nenameWalletMutation(input: RenameWalletInput) {
          const sameNameAccount = Object.values(state.accountsByAddress).find(
            (account) => account.displayName === input.displayName
          );

          if (sameNameAccount) {
            throw new Error("Wallet with that name already exists");
          }

          const account = state.accountsByAddress[input.address];

          const nextAccount = {
            ...account,
            displayName: sanitizeWalletName(input.displayName),
          };

          setState((state) => ({
            ...state,
            accountsByAddress: {
              ...state.accountsByAddress,
              [account.address]: nextAccount,
            },
          }));

          return nextAccount;
        }
        return useMutation(nenameWalletMutation);
      },

      importWallet: () => {
        async function importWalletMutation(input: ImportWalletInput) {
          const account = Object.values(state.accountsByAddress);
          const sameNameAccount = account.find(
            (account) => account.displayName === input.displayName
          );

          if (sameNameAccount) {
            throw new Error("Wallet with that name already exists");
          }

          if (!input.password.length) {
            throw new Error("Password must not be empty");
          }

          const wallet = new Wallet(input.privateKey);

          const encryptedJson = await wallet.encrypt(
            input.password,
            {},
            input.onProgress
          );

          setState((state) => ({
            ...state,
            accountsByAddress: {
              ...state.accountsByAddress,
              [wallet.address]: {
                address: wallet.address,
                encryptedJson,
                displayName: input.displayName,
              },
            },
          }));

          return wallet;
        }
        return useMutation(importWalletMutation);
      },

      unlockWallet: () => {
        async function unlockWalletMutation(input: UnlockWalletInput) {
          const account = state.accountsByAddress[input.address];

          if (!account) {
            throw new Error("Account not found");
          }

          const wallet = await Wallet.fromEncryptedJson(
            account.encryptedJson,
            input.password,
            input.onProgress
          );

          return wallet;
        }
        return useMutation(unlockWalletMutation);
      },

      deleteWallet: () => {
        async function deleteWalletMutation(input: DeleteWalletInput) {
          const sanitizedAddress = getAddress(input.address);
          const account = state.accountsByAddress[sanitizedAddress];

          // password is needed to decrypt the wallet
          // so we can delete it from the state
          await Wallet.fromEncryptedJson(
            account.encryptedJson,
            input.password,
            input.onProgress
          );

          setState((state) => ({
            ...state,
            accountsByAddress: Object.fromEntries(
              Object.entries(state.accountsByAddress).filter(
                ([key]) => key !== sanitizedAddress
              )
            ),
          }));
        }
        return useMutation(deleteWalletMutation);
      },
    };

    const queries = {
      getBalance: (params: GetBalanceParams) =>
        useQuery(
          ["eth-balance", params.address, params.provider],
          async () => {
            const provider =
              params.provider ??
              getDefaultProvider(state.selectedNetwork ?? NETWORKS.goerli, {
                alchemy: ALCHEMY_API_KEY,
              });
            const balance = await provider.getBalance(params.address);
            return formatEther(balance);
          },
          {
            enabled: isValidBIP32Address(params.address),
          }
        ),
    };

    const accountList = useMemo(
      () => Object.values(state.accountsByAddress),
      [state.accountsByAddress]
    );

    const derivedState = {
      ...state,
      accountList,
    };

    return {
      mutations,
      queries,
      state: derivedState,
    };
  });

export { WalletKeeperProvider, useWalletKeeper };
