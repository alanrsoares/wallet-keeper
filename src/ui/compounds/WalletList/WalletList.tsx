import { MinusIcon, PlusIcon } from "@heroicons/react/24/outline";

import { FC, PropsWithChildren, useState } from "react";

import { useWalletKeeper } from "~/lib/contexts/walletKeeper";
import Button from "~/ui/components/Button";

import GenerateWallet from "../GenerateWallet";
import WalletDetails from "../WalletDetails";

const WalletList: FC<PropsWithChildren> = ({ children }) => {
  const walletKeeper = useWalletKeeper();
  const [isExpanded, setIsExpanded] = useState(false);

  const walletCount = Object.keys(walletKeeper.state.accountsByAddress).length;

  if (!walletCount) {
    return <GenerateWallet />;
  }

  return (
    <div className="card bg-base-300">
      <div className="card-body grid gap-4">
        <ul className="grid gap-4">
          <h2 className="card-title flex justify-between">
            <span>
              Wallets <span className="opacity-80">({walletCount})</span>
            </span>
            <Button
              size="sm"
              shape="circle"
              onClick={() => setIsExpanded(!isExpanded)}
            >
              {isExpanded ? (
                <MinusIcon className="h-4 w-4" />
              ) : (
                <PlusIcon className="h-4 w-4" />
              )}
            </Button>
          </h2>
          {isExpanded &&
            walletKeeper.state.accountList.map(({ address, displayName }) => (
              <li key={address} className="card bg-base-200">
                <WalletDetails address={address} />
              </li>
            ))}
        </ul>
        <GenerateWallet className="bg-base-200!" />
      </div>
    </div>
  );
};

export default WalletList;
