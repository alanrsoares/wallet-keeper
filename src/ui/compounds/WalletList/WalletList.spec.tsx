import { render } from "@testing-library/react";
import { describe, expect, it } from "vitest";
import { QueryClient, QueryClientProvider } from "@tanstack/react-query";

import { WalletKeeperProvider } from "~/lib/contexts/walletKeeper";
import WalletList from "./WalletList";

const queryClient = new QueryClient();

describe("WalletList", () => {
  it("should render", () => {
    const { container } = render(
      <QueryClientProvider client={queryClient}>
        <WalletKeeperProvider>
          <WalletList />
        </WalletKeeperProvider>
      </QueryClientProvider>
    );

    // get button by test-id
    expect(container).toMatchSnapshot();
  });
});
