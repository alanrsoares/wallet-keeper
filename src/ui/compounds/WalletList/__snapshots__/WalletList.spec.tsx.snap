// Vitest Snapshot v1

exports[`WalletList > should render 1`] = `
<div>
  <button
    class="btn btn-md"
  >
    <svg
      aria-hidden="true"
      class="w-4 h-4 mr-2"
      fill="none"
      stroke="currentColor"
      stroke-width="1.5"
      viewBox="0 0 24 24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M12 4.5v15m7.5-7.5h-15"
        stroke-linecap="round"
        stroke-linejoin="round"
      />
    </svg>
    Generate new wallet
  </button>
</div>
`;
